persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateQueryHandler.java

```diff
@@ -236,81 +251,275 @@ private void authorizeByToken(ByteBuffer token, CQLStatement statement) {
           castStatement.keyspace(),
           castStatement.columnFamily());
-    } else if (statement instanceof TruncateStatement) {
-      TruncateStatement castStatement = (TruncateStatement) statement;
+    } else if (statement instanceof SchemaAlteringStatement) {
+      authorizeSchemaAlteringStatement(statement, authToken, authorization);
+    } else if (statement instanceof AuthorizationStatement) {
+      authorizeAuthorizationStatement(statement, authToken, authorization);
+    } else if (statement instanceof AuthenticationStatement) {
+      authorizeAuthenticationStatement(statement, authToken, authorization);
+    } else if (statement instanceof UseStatement) {
+      // NOOP on UseStatement since it doesn't require authorization
+      logger.debug("Skipping auth on UseStatement since it's not required");
+    } else if (statement instanceof BatchStatement) {
+      BatchStatement castStatement = (BatchStatement) statement;
+      List<ModificationStatement> statements = castStatement.getStatements();
+      for (ModificationStatement stmt : statements) {
+        authorizeModificationStatement(stmt, authToken, authorization);
+      }
+    } else {
+      logger.warn("Tried to authorize unsupported statement");
+      throw new UnsupportedOperationException(
+          "Unable to authorize statement "
+              + (statement != null ? statement.getClass().getName() : "null"));
+    }
+  }
+
+  private void authorizeModificationStatement(
+      CQLStatement statement, String authToken, AuthorizationService authorization) {
+    ModificationStatement castStatement = (ModificationStatement) statement;
+    Scope scope;
+    if (statement instanceof DeleteStatement) {
+      scope = Scope.DELETE;
+    } else {
+      scope = Scope.MODIFY;
+    }
+
+    logger.debug(
+        "preparing to authorize statement of type {} on {}.{}",
+        castStatement.getClass().toString(),
+        castStatement.keyspace(),
+        castStatement.columnFamily());
+
+    try {
+      authorization.authorizeDataWrite(
+          authToken, castStatement.keyspace(), castStatement.columnFamily(), scope);
+    } catch (io.stargate.auth.UnauthorizedException e) {
+      throw new UnauthorizedException(
+          String.format(
+              "Missing correct permission on <table %s.%s>",
+              castStatement.keyspace(), castStatement.columnFamily()));
+    }
+
+    logger.debug(
+        "authorized statement of type {} on {}.{}",
+        castStatement.getClass().toString(),
+        castStatement.keyspace(),
+        castStatement.columnFamily());
+  }
+
+  private void authorizeSchemaAlteringStatement(
+      CQLStatement statement, String authToken, AuthorizationService authorization) {
+    SchemaAlteringStatement castStatement = (SchemaAlteringStatement) statement;
+    Scope scope = null;
+    String keyspaceName = null;
+    String tableName = null;
+
+    if (statement instanceof CreateTableStatement) {
```

Comment:
Afaict, this doesn't include all schema altering statements as it doesn't include any of the statements for 2ndary indexes, materialized views, UDT, functions/aggregate or triggers. For those, the `keyspaceName, `tableName` and `scope` will remain `null`. Assuming this won't just NPE (the javadoc of `authorizeSchemaWrite` doesn't make it explicit this has to be handled), are we sure this doesn't authorize the missed operations for everyone?